File: THCS/2022_2023/TienGiang_2223/Bai3/LUCKY.py
def read_input(file_path):
    with open(file_path, 'r') as f:
        m, n = map(int, f.readline().strip().split())
        matrix = [list(map(int, f.readline().strip().split())) for _ in range(m)]
    return m, n, matrix

def get_neighbors(i, j, m, n):
    neighbors = []
    # Các chỉ số hàng xóm có thể
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]  # up, down, left, right
    for di, dj in directions:
        ni, nj = i + di, j + dj
        if 0 <= ni < m and 0 <= nj < n:
            neighbors.append((ni, nj))
    return neighbors

def count_lucky_guests(m, n, matrix):
    lucky_count = 0
    for i in range(m):
        for j in range(n):
            neighbors = get_neighbors(i, j, m, n)
            neighbor_values = [matrix[ni][nj] for ni, nj in neighbors]
            if neighbor_values:  # Đảm bảo có hàng xóm
                avg_neighbors = sum(neighbor_values) / len(neighbor_values)
                if matrix[i][j] > avg_neighbors:
                    lucky_count += 1
    return lucky_count

def write_output(file_path, result):
    with open(file_path, 'w') as f:
        f.write(str(result) + '\n')

# Thực thi các bước
input_file = 'LUCKY.INP'
output_file = 'LUCKY.OUT'

m, n, matrix = read_input(input_file)
lucky_count = count_lucky_guests(m, n, matrix)
write_output(output_file, lucky_count)

File: THCS/2022_2023/TienGiang_2223/Bai3/LUCKY.cpp
#include <iostream>
#include <vector>
#include <fstream>

using namespace std;

// Hàm đọc dữ liệu từ file
void read_input(const string &file_path, int &m, int &n, vector<vector<int>> &matrix) {
    ifstream infile(file_path);
    infile >> m >> n;
    matrix.resize(m, vector<int>(n));
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            infile >> matrix[i][j];
        }
    }
    infile.close();
}

// Hàm lấy các vị trí hàng xóm
vector<pair<int, int>> get_neighbors(int i, int j, int m, int n) {
    vector<pair<int, int>> neighbors;
    // Các chỉ số hàng xóm
    int directions[4][2] = { {-1, 0}, {1, 0}, {0, -1}, {0, 1} }; // lên, xuống, trái, phải
    for (auto& dir : directions) {
        int ni = i + dir[0];
        int nj = j + dir[1];
        if (ni >= 0 && ni < m && nj >= 0 && nj < n) {
            neighbors.push_back({ni, nj});
        }
    }
    return neighbors;
}

// Hàm đếm số người may mắn
int count_lucky_guests(int m, int n, const vector<vector<int>> &matrix) {
    int lucky_count = 0;
    for (int i = 0; i < m; ++i) {
        for (int j = 0; j < n; ++j) {
            auto neighbors = get_neighbors(i, j, m, n);
            vector<int> neighbor_values;
            for (const auto& neighbor : neighbors) {
                neighbor_values.push_back(matrix[neighbor.first][neighbor.second]);
            }
            if (!neighbor_values.empty()) { // Đảm bảo có hàng xóm
                double avg_neighbors = 0;
                for (int value : neighbor_values) {
                    avg_neighbors += value;
                }
                avg_neighbors /= neighbor_values.size();
                if (matrix[i][j] > avg_neighbors) {
                    lucky_count++;
                }
            }
        }
    }
    return lucky_count;
}

// Hàm ghi kết quả ra file
void write_output(const string &file_path, int result) {
    ofstream outfile(file_path);
    outfile << result << endl;
    outfile.close();
}

int main() {
    string input_file = "LUCKY.INP";
    string output_file = "LUCKY.OUT";

    int m, n;
    vector<vector<int>> matrix;

    read_input(input_file, m, n, matrix);
    int lucky_count = count_lucky_guests(m, n, matrix);
    write_output(output_file, lucky_count);

    return 0;
}
